### PR TITLE
Codecs

### PR DESCRIPTION
**Pull request by alissapajer**

### CODE REVIEW

**Comment by alissapajer** [comment]
Ultimately, there will be other (non-Scala) implementations of this codec, written by other applications that wish to send us `QData`. At some point we need to think through exactly how scodec is en/decoding so we can decide if we'd like to use other codecs here, either because they will be easier for others to replicate without scodec, or because they are more efficient. I haven not thought through this in detail since I wasn't sure if we needed to quite yet.

**Comment by alissapajer** [comment]
[ch62]

File: core/src/main/scala/qdata/Codec.scala [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    val codec: Codec[(A, A)] =
+      Codec.lazily(qdataCodec(version) ~ qdataCodec(version))
+
+    codec.xmap[A](
+      { case (value, meta) => qdata.makeMeta(value, meta) },
+      a => (qdata.getMetaValue(a), qdata.getMetaMeta(a)))
+  }
+
+  def isType(qType: QType): PartialFunction[A, A] = {
+    case a if qdata.tpe(a) === qType => a
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: The `===` habit dies hard. I'd actually like an anti-wart for the hot path that doesn't allow for `scalaz.Equal`. :stuck_out_tongue:

File: core/src/main/scala/qdata/Codec.scala [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+      Real.apply,
+      qdata.makeReal,
+      qdata.getReal(_).toString)
+
+  val stringCodec: Version => Codec[A] = memoize { _ =>
+    ascii32.xmap[A](qdata.makeString, qdata.getString)
+  }
+
+  val nullCodec: Version => Codec[A] = memoize { _ =>
+    provide(qdata.makeNull)
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: The `discriminated` value for this delimits it.

File: core/src/main/scala/qdata/Codec.scala [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  }
+
+  val realCodec: Version => Codec[A] =
+    exmapString[Real](
+      Real.apply,
+      qdata.makeReal,
+      qdata.getReal(_).toString)
+
+  val stringCodec: Version => Codec[A] = memoize { _ =>
+    ascii32.xmap[A](qdata.makeString, qdata.getString)
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: Oh wow good catch.

File: core/src/main/scala/qdata/Codec.scala [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+  }
+
+  ////
+
+  private def exmapString[B](parse: String => B, make: B => A, get: A => String)
+      : Version => Codec[A] = memoize { _ =>
+    ascii32.exmap[A](
+      str =>
+        Attempt.fromEither(
+          \/.fromTryCatchNonFatal(parse(str))
```
Comment (alissapajer) [inline core/src/main/scala/qdata/Codec.scala]: This all is boxing...

File: core/src/main/scala/qdata/Codec.scala [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+  utf8_32
+}
+import slamdata.Predef._
+import spire.math.Real
+
+class QDataCodec[A](qdata: QData[A]) {
+  import QType._
+
+  val longCodec: Version => Codec[A] = memoize { _ =>
+    Codec.lazily(int64.xmap[A](qdata.makeLong, qdata.getLong))
```
Comment (wemrysi) [inline core/src/main/scala/qdata/Codec.scala]: Why is this codec lazy?

File: core/src/main/scala/qdata/Codec.scala (hunk 4 above)
Comment (wemrysi) [reply core/src/main/scala/qdata/Codec.scala]: I'm fine with a `try { ... } catch { ... }` block here if that's a concern.

File: core/src/main/scala/qdata/Codec.scala (hunk 5 above)
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: Ah that's leftover. It doesn't need to be lazy, thanks!

File: core/src/main/scala/qdata/Codec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: Tests didn't catch this because, of course, they were generating ascii. I'll either find/write a unicode generator or in the very least add a individual test with unicode.

File: core/src/main/scala/qdata/Codec.scala [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        Attempt.fromOption(
+          DateTimeInterval.parse(str).map(qdata.makeInterval),
+          Err(s"Failed to parse $str as a DateTimeInterval")),
+      interval =>
+        Attempt.successful(qdata.getInterval(interval).toString))
+  }
+
+  val arrayCodec: Version => Codec[A] = memoize { version =>
+    // we encode with an extra bit prepended to each element of the encoded array
+    // instead we could encode a single "stop" flag
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply core/src/main/scala/qdata/Codec.scala]: @djspiewak I'm not sure how expensive the bit check is compared to the expense of en/decoding the array element itself. It could be measurably more efficient, I just don't know without measuring it. If this bit check is actually expensive, then I think we'd want to avoid it in the object case as well. Why were you thinking it'd be okay in the object case?

Also we have run out of space in the delimiter. We're currently using `uint4` and we'd have to move up to `unit8`, but I'm not sure that's an argument against this. It depends on what we're really trying to maximize/minimize.

**Comment by alissapajer** [comment]
@djspiewak @wemrysi ready for another look.

I've addressed everything except for possibly not encoding an extra bit before each object/array element. I'd like to have some further [discussion](https://github.com/slamdata/qdata/pull/8#discussion_r200460188) about this.

File: core/src/test/scala/qdata/TestDataGenerators.scala [hunk 7]
```diff
@@ -58,13 +58,22 @@ object TestDataGenerators {
...
     Gen.zip(genTestData(max), genTestData(max)) map { case (d1, d2) => TestData._Meta(d1, d2) })
 
-  private def listOf[A](gen: Gen[A]): Gen[List[A]] =
+  private def listOfUpTo16[A](gen: Gen[A]): Gen[List[A]] =
     for {
       n <- Gen.choose(0, 16)
       c <- Gen.listOfN[A](n, gen)
     } yield c
+
+  // FIXME generate from the space of all valid utf8-encodable characters
```
Comment (alissapajer) [inline core/src/test/scala/qdata/TestDataGenerators.scala]: Before putting this into production I'd like to spend more time on this. But it wasn't going to be trivial to solve, so for now we just choose from a couple non-ascii characters.

**Comment by wemrysi** [comment]
@alissapajer @djspiewak did you want to discuss the object/array encoding here or elsewhere? I'm ok with merging this as-is and we can revisit the encoding separately.

**Comment by alissapajer** [comment]
@wemrysi I'd be fine with discussing the encoding separately. Until this is being used we don't even have to update the version on it as we update it. 

For product version `v0.1.0` we don't need to read data from nihdb back into mimir, and no one is going to be sending us qdata-encoded bytes, so we won't be using qdata in production probably until the following version.

**Comment by djspiewak** [comment]
Yeah it's ok for us to land this and move things forward. I really do want to have a discussion about the vector encodings but that can wait.
